### PR TITLE
feat: mmdocs in bff

### LINKED ISSUE #924: feat: mm search backend
UAC
- [ ] bff activating multiple search fields (just activate)
- [ ] make sure mmdocs as search fields work

### PR DESCRIPTION
**Pull request by azayz**
closes #924 
---

- [ ] This PR references an open issue
- [ ] Test added
- [ ] Documentation adjusted

### CODE REVIEW

File: deployment/bff/app/v1/models/search.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ class SearchRequestModel(BaseRequestModel):
         default={},
         description='dictionary with filters for search results  {"tag_name" : "tag_value"}',
     )
-    query: Dict[str, ModalityModel] = Field(
+    query: List[Dict] = Field(
         default={},
         description='Nested dictionary where key can be only one of the following: ``query_text``, ``query_image``, '
```
Comment (joschkabraun) [inline deployment/bff/app/v1/models/search.py:33]: `description` needs to be updated

File: deployment/bff/app/v1/routers/helper.py [hunk 2]
```diff
@@ -15,57 +17,66 @@
...
-                    field_value.blob = None
-                    field_value.uri = file_path
-                if field_value.content is not None:
-                    data_class_kwargs[field_name_data_class] = field_value.content
+                if modalities_dict[field_name_data_class] == Image:
+                    if isinstance(field_value, str):
+                        if field_value.startswith('http'):
+                            data_class_kwargs[field_name_data_class] = field_value
+                        else:
+                            base64_decoded = base64.b64decode(
```
Comment (joschkabraun) [inline deployment/bff/app/v1/routers/helper.py:55]: we can simplify these nested `if` conditions

File: deployment/bff/app/v1/routers/helper.py [hunk 3]
```diff
@@ -15,57 +17,66 @@
...
-                    file_path = os.path.join(
-                        tmp_dir, field_name_data_class + '.' + file_ending
-                    )
-                    with open(file_path, 'wb') as f:
-                        f.write(base64_decoded)
-                    field_value.blob = None
-                    field_value.uri = file_path
-                if field_value.content is not None:
-                    data_class_kwargs[field_name_data_class] = field_value.content
+                if modalities_dict[field_name_data_class] == Image:
```
Comment (joschkabraun) [inline deployment/bff/app/v1/routers/helper.py:50]: maybe we can rephrase this as we handle only `Text` different than `Image` or `Video`

File: deployment/bff/app/v1/routers/search.py [hunk 4]
```diff
@@ -11,8 +11,13 @@
     SuggestionRequestModel,
 )
 from deployment.bff.app.v1.routers.helper import field_dict_to_mm_doc, jina_client_post
+from now.data_loading.create_dataclass import create_dataclass
 
 router = APIRouter()
+modalities_mapping = {
```
Comment (joschkabraun) [inline deployment/bff/app/v1/routers/search.py:17]: there is a function in `now.utils` for that already

File: now/data_loading/create_dataclass.py [hunk 5]
```diff
@@ -21,7 +20,9 @@ def update_dict_with_no_overwrite(dict1: Dict, dict2: Dict):
             dict1[key] = value
 
 
-def create_dataclass(user_input: UserInput):
+def create_dataclass(
+    fields: List, fields_modalities: Dict, dataset_type: DatasetTypes = None
```
Comment (joschkabraun) [inline now/data_loading/create_dataclass.py:24]: These changes require a lot of duplicate code across the code base, can we overload this function for the different scenarios?

File: deployment/bff/app/v1/routers/helper.py [hunk 6]
```diff
@@ -15,57 +17,66 @@
...
+                                    f'File type {file_ending} is not supported. '
+                                    f'Please provide a valid file type.'
+                                )
+                            file_path = os.path.join(
+                                tmp_dir, field_name_data_class + '.' + file_ending
+                            )
+                            with open(file_path, 'wb') as f:
+                                f.write(base64_decoded)
+                            field_value = file_path
+                if field_value is not None:
```
Comment (violenil) [inline deployment/bff/app/v1/routers/helper.py:78]:
```suggestion
                if field_value:
```

File: deployment/bff/app/v1/routers/helper.py [hunk 7]
```diff
@@ -15,57 +17,66 @@
...
+                if modalities_dict[field_name_data_class] == Image:
+                    if isinstance(field_value, str):
+                        if field_value.startswith('http'):
+                            data_class_kwargs[field_name_data_class] = field_value
+                        else:
+                            base64_decoded = base64.b64decode(
+                                field_value.encode('utf-8')
+                            )
+                            file_ending = filetype.guess(base64_decoded)
+                            if file_ending is None:
```
Comment (violenil) [inline deployment/bff/app/v1/routers/helper.py:59]:
```suggestion
                            if not file_ending:
```

File: deployment/playground/src/search.py [hunk 8]
```diff
@@ -171,10 +98,50 @@ def search_by_image(document: Document, jwt, filter_selection) -> DocumentArray:
...
+        'limit': top_k if top_k else params.top_k,
+        'filters': updated_dict,
+    }
+    if endpoint == 'suggestion':
+        data['text'] = query_field_values_modalities[0]['value']
+    elif endpoint == 'search':
+        data['query'] = query_field_values_modalities
+        data['create_temp_link'] = True
+        # in case the jwt is none, no jwt will be sent. This is the case when no authentication is used for that flow
+    if jwt is not None:
```
Comment (violenil) [inline deployment/playground/src/search.py:142]:
```suggestion
    if jwt:
```

File: tests/integration/bff/test_api_key.py [hunk 9]
```diff
@@ -50,7 +50,9 @@ def test_add_key(start_bff, setup_service_running, random_index_name, tmpdir):
         print('# test api keys')
         print('# search with invalid api key')
```
Comment (violenil) [inline tests/integration/bff/test_api_key.py:51]: should we convert these to comments rather than prints?

File: tests/integration/offline_flow/test_offline_flow.py [hunk 10]
```diff
@@ -30,7 +30,7 @@ def test_docarray(monkeypatch, setup_service_running, multi_modal_data):
     assert index_result == DocumentArray()
     search_result = search(
         SearchRequestModel(
-            query={'query_text': {'text': 'girl on motorbike'}},
+            query=[{'name': 'text', 'value': 'girl on motorbike', 'modality': 'text'}],
```
Comment (violenil) [inline tests/integration/offline_flow/test_offline_flow.py:33]: what exactly is name? Is it a `field_name`?

File: tests/unit/bff/v1/routers/test_search.py [hunk 11]
```diff
@@ -12,7 +12,11 @@ def test_text_search_fails_with_no_flow_running(
     with pytest.raises(ConnectionError):
         client.post(
             f'/api/v1/search-app/search',
-            json={'query': {'query_image': {'blob': base64_image_string}}},
+            json={
+                'query': [
+                    {'name': 'blob', 'value': base64_image_string, 'modality': 'image'},
```
Comment (violenil) [inline tests/unit/bff/v1/routers/test_search.py:17]: perhaps we should update our documentation with an example like this of how the request body should look like for the API.

File: tests/unit/test_dataclass.py [hunk 12]
```diff
@@ -153,7 +154,16 @@ def test_create_dataclass(
...
+    all_modalities = {}
+    all_modalities.update(user_input.index_field_candidates_to_modalities)
+    update_dict_with_no_overwrite(
+        all_modalities, user_input.filter_field_candidates_to_modalities
+    )
+    mm_doc, _ = create_dataclass(
+        user_input.index_fields + user_input.filter_fields,
+        all_modalities,
+        user_input.dataset_type,
+    )
```
Comment (violenil) [inline tests/unit/test_dataclass.py:166]: I see this logic in a few tests, should we make a function for this? May reduce duplicate code.

**Review by violenil — COMMENTED** [review]
Good work! 🚀 left some comments and suggestions.

File: tests/integration/offline_flow/test_offline_flow.py (hunk 10 above)
Comment (azayz) [reply tests/integration/offline_flow/test_offline_flow.py:33]: the idea is the user can also specify the query field name besides the value so for example you can have both description and title as text input fields so you can also specify the names

File: tests/unit/bff/v1/routers/test_search.py (hunk 11 above)
Comment (azayz) [reply tests/unit/bff/v1/routers/test_search.py:17]: I don't think we have a part of the documentation that contains this, let me know if I'm wrong. Also I think that's included in this ticket https://github.com/jina-ai/now/issues/837

File: tests/integration/offline_flow/test_offline_flow.py (hunk 10 above)
Comment (violenil) [reply tests/integration/offline_flow/test_offline_flow.py:33]: Gotcha. That's very useful!

File: tests/unit/bff/v1/routers/test_search.py (hunk 11 above)
Comment (violenil) [reply tests/unit/bff/v1/routers/test_search.py:17]: Yeah I guess we can update all of that sometime. Thanks

**Review by violenil — APPROVED** [review]
LGTM! Good work 🚀

File: deployment/bff/app/v1/routers/helper.py [hunk 13]
```diff
@@ -15,57 +17,64 @@
...
-                    base64_decoded = base64.b64decode(field_value.blob.encode('utf-8'))
-                    file_ending = filetype.guess(base64_decoded)
-                    if file_ending is None:
-                        raise ValueError(
-                            f'Could not guess file type of blob {field_value.blob}. '
-                            f'Please provide a valid file type.'
+                if modalities_dict[field_name_data_class] != Text:
+                    if isinstance(field_value, str):
+                        if field_value.startswith('http'):
+                            data_class_kwargs[field_name_data_class] = field_value
```
Comment (joschkabraun) [inline deployment/bff/app/v1/routers/helper.py:53]: these lines are the same as line 77, let's avoid this

File: deployment/bff/app/v1/routers/helper.py [hunk 14]
```diff
@@ -15,57 +17,64 @@
...
-                    )
-                    with open(file_path, 'wb') as f:
-                        f.write(base64_decoded)
-                    field_value.blob = None
-                    field_value.uri = file_path
-                if field_value.content is not None:
-                    data_class_kwargs[field_name_data_class] = field_value.content
+                if modalities_dict[field_name_data_class] != Text:
+                    if isinstance(field_value, str):
+                        if not field_value.startswith('http'):
```
Comment (joschkabraun) [inline deployment/bff/app/v1/routers/helper.py:52]: When do we need to check if `field_value` is a string? Isn't it always a string?
Also, we can combine all `if` conditions into one line without that many indents

File: deployment/bff/app/v1/routers/search.py [hunk 15]
```diff
@@ -21,14 +22,27 @@
...
+    for field in data.query:
+        fields_modalities_mapping[field['name']] = modality_string_to_docarray_typing(
+            field['modality']
+        )
+        fields_values_mapping[field['name']] = field['value']
+    data_class, field_names_to_dataclass_fields = create_dataclass(
+        fields=list(fields_modalities_mapping.keys()),
+        fields_modalities=fields_modalities_mapping,
+    )
+    query_doc = field_dict_to_mm_doc(
```
Comment (joschkabraun) [inline deployment/bff/app/v1/routers/search.py:40]: I think there are also other places where `field_dict_to_mm_doc` is used, so can you do a global search and adjust usage there as well?

File: now/data_loading/create_dataclass.py [hunk 16]
```diff
@@ -21,7 +21,12 @@ def update_dict_with_no_overwrite(dict1: Dict, dict2: Dict):
             dict1[key] = value
 
 
-def create_dataclass(user_input: UserInput):
+def create_dataclass(
+    fields: List = None,
+    fields_modalities: Dict = None,
+    dataset_type: DatasetTypes = None,
+    user_input: UserInput = None,
```
Comment (joschkabraun) [inline now/data_loading/create_dataclass.py:28]: I would either enhance the docs to make it clear to only set `user_input` or the others. Or write two different functions with different parameters to clearly separate it. If that doesn't work, maybe rename it to `create_dataclass_from_user_input`.

File: deployment/bff/app/v1/routers/search.py (hunk 15 above)
Comment (azayz) [reply deployment/bff/app/v1/routers/search.py:40]: done

File: deployment/bff/app/v1/routers/helper.py (hunk 14 above)
Comment (azayz) [reply deployment/bff/app/v1/routers/helper.py]: true, adjusted

**Review by joschkabraun — APPROVED** [review]
LGTM!
